### DISCUSSION TITLE
How to deal with "PodSecurityViolation" warnings and alerts in OKD 4.11

### DISCUSSION THREAD

**Discussion by jacksgt** [okd-project/okd#1423] [Q&A]
Hello everyone,

since upgrading to OKD 4.11 we are seeing "PodSecurityAlerts" from the monitoring stack:

> PodSecurityViolation:
> A workload (pod, deployment, deamonset, ...) was created somewhere in the cluster but it did not match the PodSecurity "restricted" profile defined by its namespace either via the cluster-wide configuration (which triggers on a "restricted" profile violations) or by the namespace local Pod Security labels. Refer to Kubernetes documentation on Pod Security Admission to learn more about these violations.

As well as these warnings (!) when creating workloads:

```
W1110 11:50:22.541336    2488 warnings.go:70] would violate PodSecurity "restricted:latest": allowPrivilegeEscalation != false (container "nginx" must set securityContext.allowPrivilegeEscalation=false), unrestricted capabilities (container "nginx" must set securityContext.capabilities.drop=["ALL"]), runAsNonRoot != true (pod or container "nginx" must set securityContext.runAsNonRoot=true), seccompProfile (pod or container "nginx" must set securityContext.seccompProfile.type to "RuntimeDefault" or "Localhost")
```

I understand that Kubernetes v1.24 (and therefore OKD 4.11) comes with a new PodSecurityAdmission mechanism which replaces the PodSecurityPolicy (removed in k8s v1.25).
However, since the pods that get created from the deployment have all the required settings (presumably injected by OpenShift's API server?), I'm unsure if any action is required.

* https://cloud.redhat.com/blog/pod-security-admission-in-openshift-4.11
* https://access.redhat.com/solutions/6976583
* https://docs.openshift.com/container-platform/4.11//authentication/understanding-and-managing-pod-security-admission.html
* https://connect.redhat.com/en/blog/important-openshift-changes-pod-security-standards

Based on my understanding so far, the PodSecurityAdmission is checking the *input objects* (Deployments, Daemonsets etc.) which are *missing* the required settings.
However, I don't want to specify all these settings on each and everyone of our applications, therefore I rely Openshift's default behavior of injecting these settings into *the pod that get created*.

Will this behavior change in the future?

**Comment by vrutkovs** [comment]
@stlaz could you help us with this?

**Comment by stlaz** [comment]
The `PodSecurityViolation` event should only appear when a Pod would be the violator - we changed the behavior in https://github.com/openshift/cluster-kube-apiserver-operator/pull/1363 but it looks like we did not change the alert description.

That means that it's possible that there are in fact some pods that might be triggering this alert, which might be a potential problem.

I recently pushed a change to https://github.com/openshift/cluster-debug-tools/pull/37 that allows you to filter your audit logs for potential violators - once you've compiled the binary, run it as `kubectl-dev_tool audit -f <path-to-audit-logs-file> --podsecurityviolations pod` to see if it's really the monitoring stack pods that are causing the issue.

As for pod controllers raising the warnings - we are aware of the problem. In fact, just yesterday, we merged a patch that's supposed to mute them in case the pod that gets created would be just fine within the given namespace's PodSecurity admission limits - https://github.com/openshift/kubernetes/pull/1393. We'll however need to test it and give it a little bit of soak time before it makes it to 4.11.

**Reply by jacksgt** [reply]
> As for pod controllers raising the warnings - we are aware of the problem. In fact, just yesterday, we merged a patch that's supposed to mute them in case the pod that gets created would be just fine within the given namespace's PodSecurity admission limits - https://github.com/openshift/kubernetes/pull/1393. We'll however need to test it and give it a little bit of soak time before it makes it to 4.11.

Thanks for this update, interesting to know!

**Reply by boldandbusted** [reply]
Apparently this is still a problem in OKD 4.12, sadly. I'm trying to install the latest HashiCorp Vault to OKD via the official Helm chart, and get this:

```shell
$ helm install --atomic --create-namespace vault-ns hashicorp/vault --set "global.openshift=true" --set "server.dev.enabled=true" --set "ui.enabled=true" --set "ingress.enabled=true"
W0421 15:15:07.302868   15989 warnings.go:70] would violate PodSecurity "restricted:latest": allowPrivilegeEscalation != false (container "sidecar-injector" must set securityContext.allowPrivilegeEscalation=false), unrestricted capabilities (container "sidecar-injector" must set securityContext.capabilities.drop=["ALL"]), runAsNonRoot != true (pod or container "sidecar-injector" must set securityContext.runAsNonRoot=true), seccompProfile (pod or container "sidecar-injector" must set securityContext.seccompProfile.type to "RuntimeDefault" or "Localhost")
W0421 15:15:07.534349   15989 warnings.go:70] would violate PodSecurity "restricted:latest": allowPrivilegeEscalation != false (container "vault" must set securityContext.allowPrivilegeEscalation=false), unrestricted capabilities (container "vault" must set securityContext.capabilities.drop=["ALL"]), runAsNonRoot != true (pod or container "vault" must set securityContext.runAsNonRoot=true), seccompProfile (pod or container "vault" must set securityContext.seccompProfile.type to "RuntimeDefault" or "Localhost")
```

OKD was installed with merely:

```shell
$ crc setup
```

Cheers.
